kernel/contracts/ZepCore.sol

```diff
@@ -3,77 +3,60 @@ pragma solidity ^0.4.18;
 import "zeppelin-solidity/contracts/math/SafeMath.sol";
 
 contract ZepCore {
   using SafeMath for uint256;
 
-  event Staked(address indexed user, KernelInstance instance, uint256 amount, uint256 total, bytes data);
-  event Unstaked(address indexed user, KernelInstance instance, uint256 amount, uint256 total, bytes data);
+  ZepToken private _token;
+  KernelRegistry private _registry;
+  KernelStakes private _stakes;
 
   uint256 public newVersionCost;
   uint256 public developerFraction;
 
-  ZepToken private _token;
-  KernelRegistry private _registry;
-  uint256 private _totalStaked;
-  mapping(address => uint256) private _instanceVouches;
-  mapping(address => mapping (address => uint256)) private _userVouches;
-
   function ZepCore(uint256 _newVersionCost, uint256 _developerFraction) public {
-    _token = new ZepToken();
     _registry = new KernelRegistry();
+    _stakes = new KernelStakes();
+    _token = new ZepToken();
     _token.transferOwnership(msg.sender);
+    
     developerFraction = _developerFraction;
     newVersionCost = _newVersionCost;
     // TODO: we need to think how we are going to manage variable costs to propose new versions
   }
 
   function register(KernelInstance instance) public {
-    registry().addInstance(instance);
-    token().transferFrom(msg.sender, this, newVersionCost);
-    token().burn(newVersionCost);
-    // TODO: should we build a burnFrom or sth like that?
+    _registry.addInstance(instance);
+    
+    // TODO: Update to burnFrom once https://github.com/OpenZeppelin/zeppelin-solidity/pull/870 is merged
+    _token.transferFrom(msg.sender, this, newVersionCost);
+    _token.burn(newVersionCost);
   }
 
-  function transferStake(KernelInstance from, KernelInstance to, uint256 amount, bytes data) public {
-    // TODO: an unstake followed by a stake call won't work
-    require(false);
+  function getInstance(string name, string version) public view returns(KernelInstance) {
+    return _registry.getInstance(name, version);
   }
 
   function stake(KernelInstance instance, uint256 amount, bytes data) public {
-    token().transferFrom(msg.sender, this, amount);
+    _token.transferFrom(msg.sender, this, amount);
     uint256 developerPayout = amount.div(developerFraction);
     require(developerPayout > 0);
-    // TODO: check how we can manage remainders in a better way
+    // TODO: Think how we can manage remainders in a better way
 
     uint256 stakedAmount = amount.sub(developerPayout);
-    _totalStaked = totalStaked().add(stakedAmount);
-    _instanceVouches[instance] = totalStakedFor(instance).add(stakedAmount);
-    _userVouches[msg.sender][instance] = _userVouches[msg.sender][instance].add(stakedAmount);
-    token().transfer(instance.developer(), developerPayout);
-
-    Staked(msg.sender, instance, stakedAmount, totalStakedFor(instance), data);
+    _stakes.stake(msg.sender, instance, stakedAmount, data);
+    _token.transfer(instance.developer(), developerPayout);
```

Comment:
what if we move the `transferFrom` just before we stake them (L49)? Just following fail-fast pattern

Reply:
Not sure I follow. `transferFrom` is the first call in the function.